File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
bool fun(TreeNode* root1, TreeNode* root2){
    // agar root wla node khali hai dono hi mei 
    if(root1==nullptr && root2==nullptr){
        return true;
    }
    // agar ek hi koi null hai ; toh symmetric toh bnefga nhi 
    if(root1==nullptr || root2==nullptr){
        return false;
    }

    // ab root ka value hi alg hpo
    if(root1->val!=root2->val){
        return false;
    }

    // ab left child of original cmprd to right child of the miirror world
    bool r1=fun(root1->left,root2->right);

    // ab right child of original compare hoga left child of the mirror world ke 
    bool r2=fun(root1->right,root2->left);

    return (r1 && r2);                  // ye return tabhi hoga jab dono true ho otherwise false aaega
}

    bool isSymmetric(TreeNode* root) {
        // Uska left child ek tree mno and dusra right child dusra tree
        return fun(root->left,root->right);
    }
};

File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public boolean fun(TreeNode root1, TreeNode root2){
        // root start wla hi null hai dono mei
        if(root1==null && root2==null){
            return true;
        }
        // agar root start wla mei se koi ek hi null hai ; toh ofc symm toh nhi hoga 
        if(root1==null || root2==null){
            return false;
        }
        // agar root present hai; par value diff hai; toh bhi false 
        if(root1.val!=root2.val){
            return false;
        }

        // left child original -> right child in mirror world 
        boolean r1=fun(root1.left,root2.right);
        // right child original -> left child of mirror world
        boolean r2=fun(root1.right,root2.left);

        return (r1 && r2);                   // agar doono true huye tabhi true return hoga 
    }

    public boolean isSymmetric(TreeNode root) {
        // main tree mein se left child ek tree bnega and rt child dusra tree 
        return fun(root.left,root.right);
    }
}